src/EditorFeatures/Core/Implementation/KeywordHighlighting/HighlightingService.cs

```diff
@@ -24,14 +24,28 @@ internal class HighlightingService : IHighlightingService
             _highlighters = highlighters.ToList();
         }
 
-        public IEnumerable<TextSpan> GetHighlights(
-             SyntaxNode root, int position, CancellationToken cancellationToken)
+        public void AddHighlights(
+             SyntaxNode root, int position, List<TextSpan> highlights, CancellationToken cancellationToken)
         {
-            return _highlighters.Where(h => h.Metadata.Language == root.Language)
-                                .Select(h => h.Value.GetHighlights(root, position, cancellationToken))
-                                .WhereNotNull()
-                                .Flatten()
-                                .Distinct();
+            using var _ = s_listPool.GetPooledObject();
+            var tempHighlights = _.Object;
+
+            foreach (var highlighter in _highlighters.Where(h => h.Metadata.Language == root.Language))
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                highlighter.Value.AddHighlights(root, position, tempHighlights, cancellationToken);
+            }
+
+            tempHighlights.Sort();
+            var lastSpan = default(TextSpan);
+            foreach (var span in tempHighlights)
+            {
+                if (span != lastSpan && !span.IsEmpty)
```

Comment:
Does this perform better than `highlights.AddRange(tempHighlights.Distinct())`? 

It seems more error prone since we're only comparing against the last one added.

Reply:
ah, since it's sorted I guess this makes it fine?

Reply:
> Does this perform better than highlights.AddRange(tempHighlights.Distinct())?

Yes.  No allocations at all here.  

> ah, since it's sorted I guess this makes it fine?

Yes.